OpenRA.Mods.Common/Traits/BotModules/UnitBuilderBotModule.cs

```diff
@@ -161,32 +147,39 @@ void BuildUnit(IBot bot, string name)
 
 		ActorInfo ChooseRandomUnitToBuild(ProductionQueue queue)
 		{
-			var buildableThings = queue.BuildableItems();
-			if (!buildableThings.Any())
+			var unitsToBuild = Info.UnitsToBuild.ToArray();
+			if (unitsToBuild.Length == 0)
 				return null;
 
-			var unit = buildableThings.Random(world.LocalRandom);
-			return HasAdequateAirUnitReloadBuildings(unit) ? unit : null;
-		}
-
-		ActorInfo ChooseUnitToBuild(ProductionQueue queue)
-		{
-			var buildableThings = queue.BuildableItems();
-			if (!buildableThings.Any())
+			var buildableThings = queue.BuildableItems().Shuffle(world.LocalRandom).ToArray();
+			if (buildableThings.Length == 0)
 				return null;
 
-			var myUnits = player.World
-				.ActorsHavingTrait<IPositionable>()
-				.Where(a => a.Owner == player)
-				.Select(a => a.Info.Name).ToList();
+			var allUnits = world.Actors.Where(a => a.Owner == player && Info.UnitsToBuild.ContainsKey(a.Info.Name) && !a.IsDead && a.IsInWorld).ToArray();
+
+			ActorInfo desiredUnit = null;
+			var desiredError = int.MaxValue;
+			foreach (var unit in buildableThings)
+			{
+				if (!Info.UnitsToBuild.ContainsKey(unit.Name) || (Info.UnitDelays != null && Info.UnitDelays.ContainsKey(unit.Name) && Info.UnitDelays[unit.Name] > world.WorldTick))
+					continue;
+
+				var unitCount = allUnits.Count(a => a.Info.Name == unit.Name);
+				if (Info.UnitLimits != null && Info.UnitLimits.ContainsKey(unit.Name) && unitCount >= Info.UnitLimits[unit.Name])
+					continue;
 
-			foreach (var unit in Info.UnitsToBuild.Shuffle(world.LocalRandom))
-				if (buildableThings.Any(b => b.Name == unit.Key))
-					if (myUnits.Count(a => a == unit.Key) * 100 < unit.Value * myUnits.Count)
-						if (HasAdequateAirUnitReloadBuildings(world.Map.Rules.Actors[unit.Key]))
-							return world.Map.Rules.Actors[unit.Key];
+				var error = allUnits.Length > 0 ? unitCount * 100 / allUnits.Length - Info.UnitsToBuild[unit.Name] : -1;
+				if (error < 0)
+					return HasAdequateAirUnitReloadBuildings(unit) ? unit : null;
+
+				if (error < desiredError)
```

Comment:
First you shuffle all units. Then to here make sure you first build the units farest off from the desired count?

Is there a need to shuffle then still? I think you loose your randomness here?

Reply:
I think we should better combine both random and unit fraction control. 

Consider this: we need medic form 40% of the infantry squad, if AI build it fully randomly, then it is possible that we will get a suqad almost all medics, which is against our design.

Reply:
You can consider the new `ChooseRandomUnitToBuild` is a combination of old `ChooseRandomUnitToBuild` and `ChooseUnitToBuild` with delay check and unit limit check.

Reply:
The random is, when units in `Info.UnitToBuild` less than the allowed fraction, AI will build unit randomly. If unit in the list is all more than allowed fraction, AI will choose unit to build to keep fraction.